torch/_inductor/codegen/cpp.py

```diff
@@ -1259,25 +1275,74 @@ def masked(mask, body, other):
         code.writeline(";")
         V.kernel.compute.splice(code)
 
+        body_code = f"{var}()"
+        body_code_vec = (
+            body_code if result.is_vec else f"at::vec::Vectorized<float>({body_code})"
+        )
         other_code = value_to_cpp(other, "float")
         other_code_vec = f"at::vec::Vectorized<float>({other_code})"
-
-        if result.is_vec:
-            type = f"decltype({var}())"
+        assert isinstance(new_mask, CppCSEVariable), new_mask
+        if new_mask.is_vec or result.is_vec:
+            type = f"decltype({body_code_vec})"
             float_mask = f"to_float_mask({new_mask})"
+            code = BracesBuffer()
+            code.writeline("[&]")
+            with V.kernel.swap_buffers(code), code.indent():
+                code.writeline(f"if (all_zero({float_mask}))")
+                with code.indent():
+                    code.writeline(f"return {other_code_vec};")
+                code.writeline("else")
+                with code.indent():
+                    code.writeline(
+                        f"return {type}::blendv({other_code_vec}, {body_code_vec}, {float_mask});"
+                    )
+            code.writeline("()")
             csevar = V.kernel.cse.generate(
                 V.kernel.compute,
-                f"{type}::blendv({other_code_vec}, {var}(), {float_mask})",
+                code,
             )
         else:
             csevar = V.kernel.cse.generate(
-                V.kernel.compute, f"{mask} ? {var}() : {other_code}"
+                V.kernel.compute, f"{mask} ? {body_code} : {other_code}"
             )
         # `result` is explicitly added to the args for correct propagation
         # of relevant itervars and vectorization status.
         csevar.update_on_args("masked", (mask, body, other, result), {})
         return csevar
 
+    @staticmethod
+    def index_expr(expr, dtype):
+        opt_ctx: OptimizationContext = get_current_node_opt_ctx()
+        assert opt_ctx and opt_ctx.dtype is not None
+        dtype = opt_ctx.dtype
+        assert dtype == torch.int32
+        assert isinstance(V.kernel, CppVecKernel)
+        index = V.kernel.rename_indexing(expr)
+        tiling_var = V.kernel.itervars[V.kernel.tiling_idx]
+        if not V.kernel.index_depends_on(index, tiling_var):
+            # if index doesn't depend on tiling_var, it is fine to use a scalar index
+            return CppOverrides.index_expr(expr, dtype)
+        if stride_at(
+            tiling_var, index
+        ).is_number and not V.kernel.index_indirect_depends_on(index, tiling_var):
```

Comment:
Why do you need the stride  to be a number? It looks like the logic would also work when the stride is dynamic.

Reply:
That's a good point. Let me see if I can refactor it a bit in a separate PR.